### ISSUE TITLE
Memory optimisation suggestion

### ISSUE THREAD

**Issue by umanayana** [aws/aws-iot-device-sdk-embedded-C#218]
The `_ClientData::writeBuf` should not be copying the caller payload into them. This produces a double buffer system which is not needed, and waste memory. Consider smaller processor with just a small amount of RAM.

* The SDK does not queue published messages thus no never a reason to have the caller payload copied over to a common buffer.
* The caller has already allocated memory for this data, and the memory will stay in scope until the publish method returns. 
* Biggest problem is if the caller is sending a large payload let say 512bytes, the caller has already allocated data for that payload. But now we need to increase the writeBuf size just for this case. 
* This buffer should be kept small only for internal message construction + what ever the large MQTT topic name length is.

If the current design is kept there will always be a second copy of the exact same data but this memory is completely useless, since there exist the exact same data else where. This is fine for 99% of embedded Linux system (since they typically have at least 64MB of ram) but this is a huge problem for baremetal where just a few bytes is a big problem.

## Here is a possible solution
 After doing this I would then replace `aws_iot_mqtt_internal_send_packet` where appropriate with `aws_iot_mqtt_internal_publish_packet`

Couldn't get the formatting to work.

`
IoT_Error_t aws_iot_mqtt_internal_send_packet(AWS_IoT_Client *pClient, size_t length, Timer *pTimer)
{
    return aws_iot_mqtt_internal_publish_packet(pClient, length, pTimer, NULL, 0);
}

IoT_Error_t aws_iot_mqtt_internal_publish_packet(AWS_IoT_Client *pClient, const size_t write_buffer_length, Timer *pTimer, uint8_t *payload, const size_t payload_size)
{
	IoT_Error_t rc;

	FUNC_ENTRY;

	if(NULL == pClient || NULL == pTimer) {
		FUNC_EXIT_RC(NULL_VALUE_ERROR);
	}

	if(write_buffer_length >= pClient->clientData.writeBufSize) {
		FUNC_EXIT_RC(MQTT_TX_BUFFER_TOO_SHORT_ERROR);
	}

#ifdef _ENABLE_THREAD_SUPPORT_
	rc = aws_iot_mqtt_client_lock_mutex(pClient, &(pClient->clientData.tls_write_mutex));
	if(SUCCESS != rc) {
		FUNC_EXIT_RC(rc);
	}
#endif

	size_t sentLen              = 0;
	size_t writebuffer_sent     = 0;

	while( writebuffer_sent < write_buffer_length && !has_timer_expired(pTimer) )
	{
		rc = pClient->networkStack.write(&(pClient->networkStack),
						 &pClient->clientData.writeBuf[writebuffer_sent],
						 (write_buffer_length - writebuffer_sent),
						 pTimer,
						 &sentLen);

		if(SUCCESS != rc) break;

		writebuffer_sent += sentLen;
	}

	size_t payload_sent = 0;

	/*Send the payload*/
	if( NULL != payload && payload_size > 0 )
	{

	    while( payload_sent < payload_size && !has_timer_expired(pTimer) )
	    {
	        rc = pClient->networkStack.write(&(pClient->networkStack),
	                         &payload[payload_sent],
	                         payload_size - payload_sent,
	                         pTimer,
	                         &sentLen);

	        if(SUCCESS != rc)  break;

	        payload_sent += sentLen;
	    }
	}

#ifdef _ENABLE_THREAD_SUPPORT_
	rc = aws_iot_mqtt_client_unlock_mutex(pClient, &(pClient->clientData.tls_write_mutex));
	if(SUCCESS != rc) {
		FUNC_EXIT_RC(rc);
	}
#endif

	const size_t total_bytes_sent     = writebuffer_sent + payload_sent;
	const size_t expect_bytes_sent    = write_buffer_length + payload_size;

	if( total_bytes_sent == expect_bytes_sent )
	{
		/* record the fact that we have successfully sent the packet */
		//countdown_sec(&c->pingTimer, c->clientData.keepAliveInterval);
		FUNC_EXIT_RC(SUCCESS);
	}

	FUNC_EXIT_RC(rc) 
}
`

**Comment by lt72** [comment]
@umanayana, thank you for the thorough analysis, we will follow up with this in a few weeks as we will be deploying some changes on our side as well. 
@gordonwang0 : please keep the above concerns in considerations for the upcoming [perf opt] work in September.

**Comment by umanayana** [comment]
I had also modified `_aws_iot_mqtt_internal_serialize_publish()` to compensate for these changes.

I am trying to optimise `AWS_IOT_MQTT_TX_BUF_LEN` so that it's the smallest possible size. What is the largest `AWS_IOT_MQTT_TX_BUF_LEN` that's needed for internal SDK usage? I ran some trivial test without studying the code and I saw the larest size is 54 bytes.

**Keep in mind** 
The `aws_iot_mqtt_publish()` method no longer copies from it's local buffer to send the payload. Thus I want to reduce `AWS_IOT_MQTT_TX_BUF_LEN` to smallest possible size needed by the SDK. 
I am only using the SDK for MQTT. I am not using shadow or Job feature

**Comment by lt72** [comment]
@umanayana one quick question: we are looking at improving the memory management as well as teh threading model by re-implementing the MQTT library to use one task to handle RX/TX, as well as re-connect and so forth... In that plan, we are considering getting rid of the RTOS less mode (aws_iot_mqtt_yield) if it is not useful to any current customers. Are you running with an RTOS or without?

**Comment by umanayana** [comment]
I am running with an RTOS, specifically FreeRTOS but I am not using the FreeRTOS provided by AWS Git repo. The FreeRTOS version I am using was downloaded directly for FreeRTOS website and I integrated IP stack, TLS and everything else required as this produces the best portability and control for our project.

I do not think you should design the library to be specifically multi-threaded, as there many cases where small devices run baremetal. There may even be cases where an RTOS environment may be so constrained that it cannot run more than just a few task maybe just 3 or 4 max. In that cases one of those threads may end up running the entire MQTT stack. From that perspective it's still as if MQTT stack is running within a baremetal environment. In my architecture only 1 thread is dedicated to running MQTT, that mean both RX/TX event occur in that thread from the stack perspective. The actual transmission across Ethernet in handle by another thread running my IP stack. 

I think the polling approach `aws_iot_mqtt_yield()` is a good idea many stack utilise this approach so that worker thread can figure out if it has work to do. Although I do not like the idea of the spin loop within the yeild method. With a good OS abstraction layer this can be done, because instead of spin looping you wait on event such RX tranmisson on the IP layer etc. But when you use a spin loop in an RTOS you are using valuable CPU time to do none valuable work. I removed the spin loop in the SDK for the yield method to prevent this from happening.

So in short I think it's good to allow for baremetal even if the environment runs RTOS because the entire stack may only ever be used within a single thread. With a good OS abstraction layer the porting layer can provide blocking for a specific time. If the porting layer doesn't want to block it just checks for the event and don't care about the time that was passed in.

I hope this helps.

**Comment by qiutongs** [comment]
Thank you for the suggestion. We are tracking this issue internally.
